Userland/Libraries/LibJS/Bytecode/Pass/AllocateRegisters.cpp

```diff
@@ -0,0 +1,173 @@
+        }
+
+        for (auto& reg : inst.read_registers()) {
+            if (max_bb < m_live_ranges[reg->index()].finish.basic_block)
+                continue;
+
+            if (block_id(basic_block) == max_bb) {
+                m_live_ranges[reg->index()].finish = { block_id(basic_block), i.offset() };
+            } else {
+                // The maximum size_t is equivalent to the end of the basic block.
+                m_live_ranges[reg->index()].finish = { max_bb, NumericLimits<size_t>::max() };
+            }
+        }
+    }
+}
+
+void AllocateRegisters::find_live_ranges(const BasicBlock* basic_block, size_t max_bb)
+{
+    if (m_live_range_path.contains_slow(basic_block))
+        return;
+
+    m_live_range_path.append(basic_block);
+    max_bb = max(block_id(basic_block), max_bb);
+
+    find_block_range(basic_block, max_bb);
+
+    const auto& children = m_executable->cfg.value().get(basic_block);
+    if (children.has_value()) {
+        for (auto& child : children.value()) {
+            find_live_ranges(child, max_bb);
+        }
+    }
+
+    m_live_range_path.take_last();
+}
+
+Vector<Register> AllocateRegisters::rename_registers()
+{
+    Vector<Register> rename;
+    for (size_t i = 0; i < m_executable->executable.number_of_registers; ++i) {
+        rename.append(Register(i));
+    }
+
+    Vector<size_t> active;
+    active.resize(m_executable->executable.number_of_registers);
+    for (auto& i : active) {
+        i = 0;
+    }
+
+    quick_sort(m_live_ranges, [](LiveRange a, LiveRange b) { return a.start < b.start; });
+
+    for (size_t i = 2; i < m_live_ranges.size(); ++i) {
+        auto& timestamp = m_live_ranges[i].start;
+
+        // Expire old intervals.
+        for (size_t j = 2; j < i; ++j) {
+            auto reg = rename[m_live_ranges[j].reg].index();
+            if (m_live_ranges[j].finish < timestamp && active[reg] == j) {
+                active[reg] = 0;
+            }
+        }
+
+        // Select a register.
+        for (size_t j = 2; j <= i; ++j) {
+            if (!active[j]) {
+                rename[m_live_ranges[i].reg] = Register(j);
+                active[j] = i;
+                break;
+            }
+        }
+    }
+
+    return rename;
+}
+
+void AllocateRegisters::apply_register_rename(Vector<Register> const& rename)
+{
+    for (auto& basic_block : m_executable->executable.basic_blocks) {
+        for (auto i = InstructionStreamIterator(basic_block.instruction_stream()); !i.at_end(); ++i) {
+            auto& inst = *i;
+            for (auto reg : inst.write_registers()) {
+                *reg = rename[reg->index()];
+            }
+
+            for (auto reg : inst.read_registers()) {
+                *reg = rename[reg->index()];
+            }
+        }
+    }
+
+    u32 number_of_registers = 0;
+    for (auto& reg : rename) {
+        max(1, 2);
+        number_of_registers = max(number_of_registers, reg.index() + 1);
```

Comment:
Doesn't seem like this has an upper bound on the register count?
From my very very fast scan through anyway?

Reply:
the upper bound is the largest register + 1. in `rename_registers` you'll see that the largest possible output register is the number of input registers - 1.

Reply:
so in the worst case, this uses only one less register? tbh I don't see how useful it is then.
if the upper bound could be specified (i.e. "use at most 8 registers or fail" to start with), it'd be closer to reality?

Reply:
Oh, I see what you mean. The JavaScript VM has infinity many registers, and has no way to spill them to memory. basically we use registers like the stack. This 'Register Allocate' doesn't assign values to architectural registers, but rather reduces the number of registers used as much as possible. Any upper bound on registers would be useless, until we JIT to x86.

Reply:
The actual upper bound is the same number of registers as the input program.

Reply:
Gotcha.
We'll rename this when we decide to add a JIT then.